scenario/state.py

```diff
@@ -755,6 +759,24 @@ def handle_path(self):
         return f"{self.owner_path or ''}/{self.data_type_name}[{self.name}]"
 
 
+@dataclasses.dataclass(frozen=True)
+class Port(_DCBase):
+    protocol: Literal["tcp", "udp", "icmp"]
+    port: Optional[int] = None
+    """The port to open. Required for TCP and UDP; not allowed for ICMP."""
+
+    def __post_init__(self):
+        port = self.port
+        if self.protocol == "icmp" and port:
+            raise StateValidationError("`port` arg not supported with `icmp` protocol")
+        elif not port:
+            raise StateValidationError(
+                f"`port` arg required with `{self.protocol}` protocol",
+            )
+        if port and not (1 <= port <= 65535):
+            raise StateValidationError(f"`port` outside bounds [1:65535], got {port}")
```

Comment:
This is obviously a matter of taste, but the use of clause guards gives us a slightly cleaner code.


```suggestion
        if not port:
            raise StateValidationError(
                f"`port` arg required with `{self.protocol}` protocol",
            )
        if self.protocol == "icmp":
            raise StateValidationError("`port` arg not supported with `icmp` protocol")

        if not (1 <= port <= 65535):
            raise StateValidationError(f"`port` outside bounds [1:65535], got {port}")
```

Reply:
you're right, it deserved a cleanup, but I ended up doing it slightly differently.